presto-main/src/main/java/com/facebook/presto/type/TypeRegistry.java

```diff
@@ -237,7 +237,52 @@ private static boolean isCovariantParameterPosition(String firstTypeBase, int po
         // Other methods should reference these two functions instead of hand-code new rules.
 
         // if we ever introduce contravariant, this function should be changed to return an enumeration: INVARIANT, COVARIANT, CONTRAVARIANT
-        return firstTypeBase.equals(StandardTypes.ARRAY);
+        return firstTypeBase.equals(StandardTypes.ARRAY) || firstTypeBase.equals(StandardTypes.MAP);
+    }
+
+    /*
+     * Return true if actualType can be coerced to expectedType AND they are both binary compatible (so it's only type coercion)
+     */
+    public static boolean isTypeOnlyCoercion(TypeSignature actualType, TypeSignature expectedType)
+    {
+        if (!canCoerce(actualType, expectedType)) {
+            return false;
+        }
+
+        if (actualType.equals(expectedType)) {
+            return true;
+        }
+        else if (actualType.getBase().equals(StandardTypes.VARCHAR) && expectedType.getBase().equals(StandardTypes.VARCHAR)) {
```

Comment:
In line 248 we first ensure that we `canCoerce(actualType, expectedType)`. In case of `isTypeOnleCoercion(varchar(40), varchar(5))` this check in line 248 would return false. In other words, we know that we can coerce `actualType` into `expectedType` and now we check that if both are some `varchar` types, then this is type only coercion.